### ISSUE TITLE
ComDB2 Postgres SQL Equivalents

### ISSUE THREAD

**Issue by saatviks** [bloomberg/comdb2#554]
What would be the equivalents of the following PostgresSQL constructs in ComDB2. I've consulted the documentation and could not find it(These are related to `CREATE TABLE` statements):

1. `DEFAULT CURRENT_TIMESTAMP`: Setting a timestamp(or any) column default value
2. `UNIQUE`: Applying a unique constraint
3. `o_carrier_id int NULL DEFAULT NULL`

**Comment by nirbhayc** [comment]
```
create table t1(c1 datetime default current_timestamp, c2 int unique, c3 int null, c4 decimal32) $$
```
I think you'd need to quite the decimal32 values in your DMLs.

**Comment by riverszhang89** [comment]
1. 
```SQL
CREATE TABLE t1 (t DATETIME DEFAULT CURRENT_TIMESTAMP)
```

2. 
```SQL
CREATE UNIQUE INDEX t1_t ON t1(t)
```

3. 

```SQL
-- Making the column nullable implies the default value of the column is NULL. 
-- You don't have to specify that again unless you'd like to choose a different default value.
CREATE TABLE t2 (i INT NULL)
```

4.
You'd need to quote the decimal value to indicate it is fixed-precision. If you use the JDBC driver, we recommend you use `PreparedStatement.setBigDecimal()` to bind a decimal value to a parameterized query.

**Comment by saatviks** [comment]
Thanks for your responses - May I keep this question open for a little while longer so I can ask similar questions as I go about the assignment.

**Comment by saatviks** [comment]
I'm getting the following error: `[UPDATE OORDER SET O_CARRIER_ID= 8 AND O_D_ID=2101 AND O_W_ID = 12] failed with rc 202 transaction too big` - The corresponding SQL Query is
`UPDATE OORDER SET O_CARRIER_ID= 8 AND O_D_ID=2101 AND O_W_ID = 12`. Any suggestion on how to resolve this?

**Comment by saatviks** [comment]
Is it sufficient to simply set `maxosqltransfer` to an arbitrarily larger value like 500000 in `$COMDB2_ROOT/etc/cdb2/config/comdb2.lrl`?

**Comment by riverszhang89** [comment]
Yes. `maxosqltransfer` should do the trick. The error log can be ignored.
You're very welcome to keep the thread open and ask questions here. We're happy to help.

**Comment by saatviks** [comment]
I've got ComDB2 running the TPCC benchmark on OLTPBench with submitting a single transaction at a time. But when trying to run concurrently by submitting multiple transactions at a time, there are errors in the results returned. Is there any setting I need to tune for Concurrency?

**Comment by Blade-Lee** [comment]
Currently I'm using conversions below from Postgres to Comdb2 (in `CREATE TABLE` statements):
1. `double precision` -> `double`
2. `smallint` -> `short`
3. `bigint` -> `longlong`

May I ask am I doing it correctly? 

Besides, what are the equivalents of Postgres `serial` type and `text` type in comdb2?

**Comment by riverszhang89** [comment]
| Comdb2 | Postgres |
|-----------|-----------|
| double    | double precision |
| smallint   | smallint |
| bigint      | bigint    |
| N.A.       | serial |
| text        | text |

**Comment by Blade-Lee** [comment]
how about `auto_increment`? such as `ipb_id int NOT NULL auto_increment, ` in a `CREATE TABLE` statement

**Comment by roborivers** [comment]
Auto increment is not supported yet.

**Comment by Blade-Lee** [comment]
How about the `UNIQUE` issue mentioned above?

**Comment by riverszhang89** [comment]
I'd recommend that you create a primary index on a set of columns that can uniquely identify a row.
`UNIQUE` and `DEFAULT` are contradictory: It is impossible to enforce uniqueness if a column has a default value.

**Comment by Blade-Lee** [comment]
After using CREATE UNIQUE INDEX, we addressed the `UNIQUE` issue.

However, we met another issue when using:

```
CREATE TABLE ipblocks (
  ...
  ipb_timestamp varchar(14) NOT NULL DEFAULT '\0\0\0\0\0\0\0\0\0\0\0\0\0\0',
  ...
);
```
there is an error for this stmt:

```
2017/12/13 04:32:34 [ERROR] init_default_value: dbstore for ipb_timestamp is invalid
2017/12/13 04:32:34 [ERROR] add_table_to_environment: add_cmacc_stmt failed
```

May I ask how should we correctly rephrase the stmt? and what is `add_cmacc_stmt`?

**Comment by riverszhang89** [comment]
You'd use:

```SQL
CREATE TABLE ipblocks (ipb_timestamp varchar(14) NOT NULL DEFAULT "" WITH dbpad=0)
```
(See "Column constraint" under [the "CREATE TABLE" section](https://bloomberg.github.io/comdb2/sql.html#create-table))

**Comment by QinyuHu** [comment]
I am looking for a DB2 version of Date. I checked this https://bloomberg.github.io/comdb2/datetime.html and not sure I found an equivalent SQL version. Did I miss something? The original is   `o_orderdate       DATE NOT NULL,` from TPCH benchmark.

**Comment by adizaimi** [comment]
I think comdb2's datetime type will behave similar to DATE for your purpose.
